From 1b3d6e6f3ca6f0210b103960f8221c57b65a6b00 Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Date: Thu, 10 Dec 2020 14:57:57 +0200
Subject: [PATCH 12/78] sram: Add LLCE sram compatible

This will initialize SRAM memory and reset ECC.

Issue: ALB-5770
Signed-off-by: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
---
 drivers/misc/sram.c | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/drivers/misc/sram.c b/drivers/misc/sram.c
index f30448bf3a63..d09a10158067 100644
--- a/drivers/misc/sram.c
+++ b/drivers/misc/sram.c
@@ -317,7 +317,7 @@ static int sram_reserve_regions(struct sram_dev *sram, struct resource *res)
 	return ret;
 }
 
-static int atmel_securam_wait(void)
+static int atmel_securam_wait(struct platform_device *pdev)
 {
 	struct regmap *regmap;
 	u32 val;
@@ -331,9 +331,19 @@ static int atmel_securam_wait(void)
 					10000, 500000);
 }
 
+static int llce_init_sram(struct platform_device *pdev)
+{
+	struct sram_dev *sram = platform_get_drvdata(pdev);
+	size_t size = gen_pool_size(sram->pool);
+
+	memset_io((void __iomem *)sram->virt_base, 0, size);
+	return 0;
+}
+
 static const struct of_device_id sram_dt_ids[] = {
 	{ .compatible = "mmio-sram" },
 	{ .compatible = "atmel,sama5d2-securam", .data = atmel_securam_wait },
+	{ .compatible = "nxp,s32g274a-llce-sram", .data = llce_init_sram },
 	{}
 };
 
@@ -343,7 +353,7 @@ static int sram_probe(struct platform_device *pdev)
 	struct resource *res;
 	size_t size;
 	int ret;
-	int (*init_func)(void);
+	int (*init_func)(struct platform_device *pdev);
 
 	sram = devm_kzalloc(&pdev->dev, sizeof(*sram), GFP_KERNEL);
 	if (!sram)
@@ -390,7 +400,7 @@ static int sram_probe(struct platform_device *pdev)
 
 	init_func = of_device_get_match_data(&pdev->dev);
 	if (init_func) {
-		ret = init_func();
+		ret = init_func(pdev);
 		if (ret)
 			goto err_free_partitions;
 	}
-- 
2.25.1

